app/src/main/java/io/gnosis/safe/ui/settings/owner/ledger/LedgerController.kt

```diff
@@ -0,0 +1,275 @@
+import android.os.ParcelUuid
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.Fragment
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.parseGetAddress
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.splitPath
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.unwrapADPU
+import io.gnosis.safe.ui.settings.owner.ledger.LedgerWrapper.wrapADPU
+import io.gnosis.safe.ui.settings.owner.ledger.ble.ConnectionEventListener
+import io.gnosis.safe.ui.settings.owner.ledger.ble.ConnectionManager
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.conflate
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.toHexString
+import timber.log.Timber
+import java.util.*
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.suspendCoroutine
+
+class LedgerController(val context: Context) {
+
+    var connectedDevice: BluetoothDevice? = null
+        private set
+
+    private var deviceConnectedCallback: DeviceConnectedCallback? = null
+    private var addressContinuation: Continuation<Solidity.Address>? = null
+
+    var writeCharacteristic: BluetoothGattCharacteristic? = null
+    var notifyCharacteristic: BluetoothGattCharacteristic? = null
+
+    fun loadDeviceCharacteristics() {
+        val characteristic = connectedDevice?.let {
+            ConnectionManager.servicesOnDevice(it)?.flatMap { service ->
+                service.characteristics ?: listOf()
+            }
+        } ?: listOf()
+        writeCharacteristic = characteristic.find { it.uuid == UUID.fromString("13d63400-2c97-0004-0002-4c6564676572") }
+        notifyCharacteristic = characteristic.find { it.uuid == UUID.fromString("13d63400-2c97-0004-0001-4c6564676572") }
+    }
+
+
+    private val bluetoothAdapter: BluetoothAdapter by lazy {
+        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+        bluetoothManager.adapter
+    }
+
+    private val bleScanner by lazy {
+        bluetoothAdapter.bluetoothLeScanner
+    }
+
+    private val connectionEventListener by lazy {
+
+        ConnectionEventListener().apply {
+
+            onConnectionSetupComplete = { gatt ->
+                Timber.d("Connected to ${gatt.device.name}")
+                connectedDevice = gatt.device
+                deviceConnectedCallback?.onDeviceConnected(connectedDevice!!)
+            }
+
+            onDisconnect = {
+                connectedDevice = null
+            }
+
+            onCharacteristicRead = { _, characteristic -> }
+
+            onCharacteristicWrite = { _, characteristic -> }
+
+            onMtuChanged = { _, mtu -> }
+
+            onCharacteristicChanged = { _, characteristic ->
+                val unwrappedResponse = unwrapADPU(characteristic.value)
+                val address = parseGetAddress(unwrappedResponse)
+                Timber.d(address)
```

Comment:
I don't mind having debug log statements. But when seeing them in logcat, it would helpful to have some context to this address. For example: 
```
Timber.d("onCharacteristicChanged() | Parsed address: $address")
```